### DISCUSSION TITLE
Champion: relax ordering constraints around `ref` and `partial` modifiers on type declarations

### DISCUSSION THREAD

**Discussion by agocke** [dotnet/csharplang#8965] [Language Ideas]
Currently, `partial` must appear directly before `struct`, `class`, or another type declaration keyword. If the type is a `ref` struct, `ref` must appear immediately before `partial` or `struct`. It seems likely that various other keywords could be used to disambiguate these contextual modifiers and allow us to relax the constraints on where `partial` and `ref` can appear in the modifier list.

- [ ] Proposal added
- [x] Discussed in LDM
- [ ] Decision in LDM
- [ ] Finalized (done, rejected, inactive)
- [ ] Spec'ed

See https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-12-04.md

**Comment by CyrusNajmabadi** [comment]
Note: i've looked at https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-12-04.md, but i can't find where in it it refers to this issue.  Can you link/copy the appropriate section?

Thanks!

**Comment by Joe4evr** [comment]
Can restrictions around `partial` just be relaxed entirely? I'd like it if `partial abstract class` worked just the same as `abstract partial class`.

**Comment by Joe4evr** [comment]
> See https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-12-04.md

So that link does not talk about `partial` *anywhere* (as @CyrusNajmabadi noted), so I did a quick search for the paper trail myself:
* https://github.com/dotnet/csharplang/blob/main/meetings/2017/LDM-2017-09-27.md (which was before this issue)
* https://github.com/dotnet/csharplang/blob/main/meetings/2018/LDM-2018-04-04.md (where it was approved)
* https://github.com/dotnet/csharplang/blob/main/meetings/2019/LDM-2019-09-16.md (where it was initially slated for 9.0)
* https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-09-09.md (where it was punted to later)

**Comment by RikkiGibson** [comment]
Opened a new champion issue #9804 and have updated the proposal for this.

**Reply by julealgon** [reply]
@RikkiGibson wouldn't it be preferred to create a separate discussion in cases like this? Otherwise, people who already voted on it might not see the (potentially drastic) change in scope and might not have the chance to reconsider their votes.

The only reason I bring this up is that votes are used in this repo in a few cases to prioritize issues based on how popular they are etc.

(I can't see the edits so disregard the above if it is not that significant in the first place).

**Reply by RikkiGibson** [reply]
In this case, I really think no. The change in scope is not drastic. It's just applying the same change to all the declaration forms that are doing this one thing--and there are more of those forms in 2025 than there were in 2017. The biggest obstacle to doing this feature, in my view, is that it is relatively small in both implementation effort and impact, and there has always been something else more important.

**Comment by TahirAhmadov** [comment]
Meh. 
Frankly, I think the order of modifiers should be mandated by the language, I think there is value to uniformity from the perspective of reading the code. 
Editors (like VS) can be improved a little to just allow typing them in any order, with immediate error highlighting/underlining, and quick fixer to put them in order.
Something similar already happens with fields, there it's not an error, but VS suggests to fix the order. Perhaps both that scenario and this scenario can meet in the middle - both should be warnings (at the end of the day, the code can compile). But I do think uniform order of modifiers across all C# code is a benefit which we shouldn't discount.

**Reply by RikkiGibson** [reply]
The feature [exists in the platform](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0036), and it should be straightforward to modify your project to use it if you wish.

**Reply by TahirAhmadov** [reply]
I know it does, but I don't think we should make it more flexible here. Just looking at the proposal you created (linked above), it's hard to understand what's going on when reading the code quickly. This will inevitably lead to different teams adopting various "styles", making it more difficult to collaborate - new SWE joins a project, has to re-adjust to the style in effect; if he works on multiple projects, he has to keep "getting used to" each project.
The question is, what is the benefit of this? Is it really that hard to type the modifiers in the same order? If it's just ease of typing, like I said, it's more of an editor improvement, not a language change.
PS. If anything, that IDE0036 should default to warning by default - we don't really have to change the language at this point.
PPS. Basically, do nothing with the language; improve editors to simplify typing of type modifiers; change the default level of IDE0036 to warning - this would be the way forward here, IMO. Just my 2 cents.

**Reply by Richiban** [reply]
> Frankly, I think the order of modifiers should be mandated by the language

I'm not sure I agree; I think that's an issue for linters / formatters. In much the same way that `private` is not required for class members, but almost all formatters add it in. It's definitely a *code style* issue; I'm not a fan of languages that literally prevent you from running the application because of some trivial thing like this. 

For example: in some languages it's literally a compiler error to have an unused variable and it's such a PITA because you just comment out some code whilst debugging but that leaves you with unused variables which you now have to delete if you want to run the program.

**Comment by LWChris** [comment]
I can't explain why, but it feels like the access modifier (`public`, `private`, `internal`, etc.) should come first no matter what.

`private partial X` feels more logical to me than `partial private X`.

For `ref` vs `partial` I have less bias, but in general, I prefer a single mandatory order for practical reasons; if you are allowed to choose the order, and possibly set a preferred order for a formatter, then this can cause "format on save" to create unintentional edit wars from tooling, or worse, cause conflicts in VCS. The less variability the order has, the better.

It's the very reason my whole team has agreed on using XAML Styler and ResxManager "alphabetical sorting" options, because that guarantees that "no matter what you do, when save, it's consistent with what everybody else would've committed".

In fact, I feel the option to define an order for code formatting is less about

> Cool, you have some choices here, tell me what you like best.

but rather 

> The order isn't mandatory, so something could introduce variance in your code.
> Here, let me fix that unfortunate circumstance, so you don't have to deal with the fallout that could cause.

If someone wants to have a VS[Code] plugin that _displays_ them in a different order (or replaces some of them with an icon), that'd be cool. But the order in the actual source code file should be as determinisitic as possible, IMHO.

**Reply by CyrusNajmabadi** [reply]
> I can't explain why, but it feels like the access modifier (public, private, internal, etc.) should come first no matter what.

THat is already not at all the case in the language.

> private partial X feels more logical to me than partial private X.

This has nothing to do with partial's relationship with accessibility modifiers, and is about instead it's requirement of being attached to the type modifier.  

for example, i cannot say `partial static class`, i *must* say `static partial class`. 

>But the order in the actual source code file should be as determinisitic as possible, IMHO.

The language does not care here.  That's a linting choice for a person/team.  The lang has allowed modifiers in any order (outside of this case) since the beginning. This just makes this modifier consistent with that.

**Reply by LWChris** [reply]
Oh, I work in C# for 10 years now and never know. It sure felt like there was only a handful of combinations where permutations were valid. Maybe I was thinking too much about `public static implicit operator X`. I realize now that this is 4 _keywords_ where only 2 out of 24 thinkable permutations are valid, but the modifiers `public` and `static` are exactly what can be re-ordered into `static public`.

Anyway, in this case I agree that if `ref` is an outlier then relaxing makes sense.
